cloudinit/distros/tests/test_networking.py

```diff
@@ -74,6 +77,34 @@ def test_returns_true_if_device_is_physical(self, sys_class_net):
         assert LinuxNetworking().is_physical(devname)
 
 
+class TestBSDNetworkingTrySetLinkUp:
+    def test_raises_notimplementederror(self):
+        with pytest.raises(NotImplementedError):
+            BSDNetworking().try_set_link_up("eth0")
+
+
+@mock.patch("cloudinit.net.is_up")
+@mock.patch("cloudinit.distros.networking.subp.subp")
+class TestLinuxNetworkingTrySetLinkUp:
+    def test_calls_subp_return_true(self, m_subp, m_is_up):
+        devname = "eth0"
+        m_is_up.return_value = True
+        is_success = LinuxNetworking().try_set_link_up(devname)
+
+        assert (mock.call(['ip', 'link', 'set', devname, 'up']) ==
+                m_subp.call_args_list[-1])
```

Comment:
@OddBloke When I set autospec=true for subp, mocking fails with "TypeError: name must be a str, not a MagicMock". Tried a few things but wasn't able to figure out the issue. So followed the hacking doc to test with non-autospecced mock here instead.

Reply:
Excellent, thanks for letting me know. `subp` _is_ actually a corner case. We have a mechanism to mock it universally (so that we don't ever accidentally run arbitrary commands on developer systems); _that_ mechanism autospecs its mock, meaning that when you autospec it here things go sideways.

(As an aside, you're the second person to hit this recently, so it's probably worth a mention in the HACKING doc.)